optimized memcpy

dweiller opened the pull request:
This PR does a few things to improve `memcpy` performance:
  - the implementation is generic, whereas master (since f38d7a92cc) is adapted from an implementation designed for x86_64 avx2 machines
  - reduce the number of branches to handle small copy lengths
  - reduce the size of generated machine code

Here is a graph or the performance of master (commit 5cfcb01503) and b7a887f0fb in a microbenchmark. The benchmark times `@memcpy` for a specific length and alignment (mod 32) of source and destination 100 000 times in a loop; this is done for all 256 combinations of source and destination alignment and the average time per iteration across all alignment combinations is the reported result.

![hot-loop](https://github.com/user-attachments/assets/def96218-d5d3-47bc-82ac-e6c204a2cab6)

_Note that this benchmark is not going to be particularly realistic for most circumstances as branch predictors will be perfectly trained._

In general we want to focus on optimizing small length copies as small lengths are the most frequent—some reported distributions can be found at [here](https://github.com/nadavrot/memset_benchmark) and [this google paper](https://storage.googleapis.com/gweb-research2023-media/pubtools/pdf/4f7c3da72d557ed418828823a8e59942859d677f.pdf) (LLVM uses these for benchmarking). For small lengths, the above graph indicates between a modest and significant (up to around 3x) improvement, depending on the length.

Here is a graph showing performance across the distributions from the linked paper linked above (taken from the LLVM repo).

![distribution](https://github.com/user-attachments/assets/3fef6022-4834-400c-96df-b6df80cb9d61)

While this PR is not focussed on `memmove` I have also changed the `ReleaseSmall` implementation of `memmove`, as the new (since f38d7a92cc) `memmove` implementation is not handled well by LLVM in `ReleaseSmall` mode.

For code size of memcpy:
| optimize mode| cpu | master 5cfcb01503 (B) | b7a887f0fb (B) |
| --------------------- | ----- | ----------------------------:   |     --------------:     |
| ReleaseFast | x86_64 | 1402 | 351 |
| ReleaseFast | x86_64_v2 | 1402 | 351 |
| ReleaseFast | x86_64_v3 | 1348 | 470 |
| ReleaseFast | x86_64_v4 | 1348 | 470 |
| ReleaseSmall | x86_64 | 1962 | 29 |
| ReleaseSmall | x86_64_v2 | 1962 | 29 |
| ReleaseSmall | x86_64_v3 | 1400 | 29 |
| ReleaseSmall | x86_64_v4 | 1400 | 29 |

_Note that the above column for master contains the size of the `memmove` implementation, the `memcpy` function on master compiles to 10 bytes of stack shuffling and a jump to `memmove`._

The `ReleaseSmall` size of `memmove` for b7a887f0fb is 58 bytes for the cpus above.

I have only checked the performance of this change on my (x86_64, avx2, znver1) machine—it would be good to check on other architectures as well if someone with one would like to test it. The benchmarks used can be found [here](https://github.com/dweiller/zig-lib-bench)—the `average` and `distrib` ones were used to produce the above charts (check out the `tools` directory for data generation scripts).

A few other notes:
  - `memcpy` and `memmove` no longer share implementations
  - the copy loops are not manually unrolled (LLVM will unroll them on various targets).
  - testing manual unrolling on my machine has not indicated a significant difference in the distribution benchmark, but does show regressions in the hot loop benchmark
  - I have not benchmarked `ReleaseSmall` performance
  - For x86_64 machines that have ERMSB it is likely best to use `rep movsb` for copies over several hundred bytes (I don't have a suitable machine to test this)
  - For x86_64 machines that have FSRM it may be better to `rep movsb` for all/most cases (I don't have a suitable machine to test this)
  - When taking advantage of ERMSB and FSRM there would be the option of runtime detection—detection should have negligible performance impact as the branch should always be predicted correctly after the first few calls
  - recent arm (e.g. armv8.8) machines may want to use the instructions from the mops feature (similar to `rep movsb` from x86_64)
  - there may exist targets that would prefer to have a branch to select different instruction when aligned loads can be done in the copy loop

Other stuff that can be done (either before merging or as follow up issues):
  - [ ] check performance on other architectures and CPUs (would need other people to run benchmarks on their machines, or put some new benchmarks in gotta-go-fast)
  - [ ] investigate `memmove` performance and whether it makes sense to merge/share code with `memcpy` again
  - [x] ~~see if there is a reasonable way to do aligned vector moves for misaligned case (doubtful without inline assembly)~~ pretty sure this can't be done without `@select` taking a runtime mask or using inline assembly
  - [x] benchmark copies with random sizes

Rexicon226 commented:
Nice work! Could you share the benchmark script? I am really curious to see how those aligned loads/stores are emitted.

dweiller commented:
> Nice work! Could you share the benchmark script? I am really curious to see how those aligned loads/stores are emitted.

The benchmark script is just a pretty simple loop copying from one buffer to the other a bunch of times for each source alignment offset. I haven't done anything clever in the micro benchmark. The reason aligned vector moves are emitted is the use of `@ptrCast` to get pointers of type `[*]CopyType`/`[*]const CopyType` in `lib/compiler_rt/memcpy.zig`. 

<details>
<summary> Here's the micro benchmark (doesn't look like I can colour it inside the fold):</summary>
<pre>
pub fn main() !void {
    const allocator = std.heap.page_allocator;

    const args = try std.process.argsAlloc(allocator);
    defer allocator.free(args);
    if (args.len > 3 or args.len < 2) return error.BadArgs;

    const iterations = try std.fmt.parseInt(usize, args[1], 10);

    const copy_len = if (args.len >= 3)
        try std.fmt.parseInt(usize, args[2], 10)
    else
        std.mem.page_size;

    std.debug.print("copying blocks of size {d}\n", .{std.fmt.fmtIntSizeBin(copy_len)});

    const alignment = if (std.simd.suggestVectorLength(u8)) |vec_size|
        @alignOf(@Type(.{ .Vector = .{ .child = u8, .len = vec_size } }))
    else
        @alignOf(usize);

    var times: [alignment]u64 = undefined;

    // preheat
    try iteration(allocator, alignment, copy_len, alignment - 2, &times[0], iterations);

    for (0..alignment) |s_offset| {
        try iteration(allocator, alignment, copy_len, s_offset, &times[s_offset], iterations);
        std.debug.print(
            "s_offset: {d}, time: {}\n",
            .{
                s_offset,
                std.fmt.fmtDuration(times[s_offset]),
            },
        );
    }

    var avg: u128 = 0;
    for (times) |t| {
        avg += t;
    }
    avg /= times.len;

    const final: u64 = @intCast(avg);

    const throughput: u64 = @intCast((@as(u128, copy_len) * iterations * std.time.ns_per_s) / final);

    std.debug.print("average time: {}, throughput: {d:.2}\n", .{
        std.fmt.fmtDuration(final),
        std.fmt.fmtIntSizeBin(throughput),
    });
}

inline fn iteration(
    allocator: Allocator,
    comptime alignment: comptime_int,
    copy_len: usize,
    s_offset: usize,
    p_time: *u64,
    iterations: usize,
) !void {
    const src = try allocator.alignedAlloc(u8, alignment, copy_len + s_offset);
    defer allocator.free(src);

    const dest = try allocator.alignedAlloc(u8, alignment, copy_len);
    defer allocator.free(dest);

    var timer = try std.time.Timer.start();

    var i: usize = 0;
    while (i < iterations) : (i += 1) {
        @memcpy(dest[0..copy_len], src[s_offset..][0..copy_len]);
    }

    const time = timer.read();

    p_time.* = time;
}

const std = @import("std");
const Allocator = std.mem.Allocator;
</pre>
</details>

Rexicon226 commented:
Interesting. I have attempted to re-produce this and I see 2, maybe 3 issues.

1. This generates only like 1 pair of aligned loads, when you could be fitting in way more. See: https://zig.godbolt.org/z/YxEPvssMa

2. I simply cannot replicate your results using the benchmark script you provided. On my AVX2 machine, I see a consistent 30% worse perf on your function. I did these measurements by simply compiling a version of LLVM Zig with both compiler_rt versions. Furthermore, I wrote my own benchmark script, and also cannot replicate it. However there I was measuring how many cycles it takes to copy X amount of bytes. Very comparable results to the benchmark script your provided. 
Note: all data collected was normalized and the mean was found with a rolling average of 50 data points over 10kb total range. 

3. If you remember our discussion about this before, my personal `memcpy` function was faster, and that still seems to be the case. Please do not interpret the before sentence as criticism or containing any mal-intent, I am simply stating the facts I observe. I believe it is because my version contains more aligned vector loads squeezed in and does a better job at managing pre-fetching. 

I recommend you use something like:
```zig
dest[0..v].* = @as(*const @Vector(v, u8), @alignCast(@ptrCast(aligned_src[offset..][0..v]))).*;
```
to generate your aligned vector loads, 😉

dweiller commented:
> 1\. This generates only like 1 pair of aligned loads, when you could be fitting in way more. See: https://zig.godbolt.org/z/YxEPvssMa
>
> 3\. ... I believe it is because my version contains more aligned vector loads squeezed in and does a better job at managing pre-fetching.

I am not convinced this is the case. It is trivial to produce a longer sequence of aligned move operations by unrolling the loop - though this does neccesitate extra cleanup code - and when I did this in the past it did not have much impact on performance. Unless I get benchmark results clearly showing a win that seems generalizable I wouldn't want to do it; forcing this sort of unrolling seems more likely to trip up the optimizer and may only positively impact performance on the machine you run the benchmark on while degrading performance on other machines.

> 2\. I simply cannot replicate your results using the benchmark script you provided.

If you let me know precisely how you compiled/ran it I will try on my machine as well. Particular things of interest would be how many iterations you did, the copy length, target cpu features and optimize mode.

> 2\. I see a consistent 30% worse perf on your function. I did these measurements by simply compiling a version of LLVM Zig with both compiler_rt versions.

Can you be more specific about what you did when you saw worse performance relative to master so I can investigate? I have also used compiling the zig compiler itself as a test and saw a minor (not really sure it was outside uncertainty) improvement in compile time.

> 2\. Furthermore, I wrote my own benchmark script, and also cannot replicate it. However there I was measuring how many cycles it takes to copy X amount of bytes. Very comparable results to the benchmark script your provided.

Can you post your benchmark? I'm not sure what you mean by "...and also cannot replicate it" along with "Very comparable results to the benchmark script your provided." - those read as contradictory statements to me, so I must be misinterpretting something.

dweiller commented:
Looks like this somehow broke a bunch of non-x86_64 targets in the linux CI. I wonder if on those targets LLVM is making memcpy produce a call to itself...

dweiller commented:
> I recommend you use something like:
> 
> ```zig
> dest[0..v].* = @as(*const @Vector(v, u8), @alignCast(@ptrCast(aligned_src[offset..][0..v]))).*;
> ```
> 
> to generate your aligned vector loads, 😉

I assume you mean `@as(*Vector(V, u8), @alignCast(@ptrCast(dest[0..v]))).* = ...`? Otherwise that is going to produce an unaligned store. Using this particular snippet to perform the case (casting the dest as well) doesn't make a difference.

Jarred-Sumner commented:
I suspect zig’s memcpy could get a whole lot faster than 30%-ish. I think it’s hard to rule out measurement noise without also having a complete benchmark and on larger amounts of data. There must already be comprehensive memcpy benchmarks online somewhere you could copy from

dweiller commented:
> I suspect zig’s memcpy could get a whole lot faster than 30%-ish.

For small sizes this is certainly true, but for large copies I'm a bit skeptical, at least not without using inline asm with something like `rep movsb`, but that opens up a bit of a can of worms.

dweiller commented:
I've made a bunch of improvements yielding both better performance and smaller code size. The table in the top post has been updated.

dweiller commented:
> Edit: Not sure what the problem with riscv64-linux is in the CI - I've disassembled it and memcpy and is not doing a recursive call.

This is wrong a recursive `jalr` instructions are generated.

dweiller commented:
The windows CI failure is due ` CMake Error: CMake was unable to find a build program corresponding to "Ninja".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.` Not sure why that it happening.

RossComputerGuy commented:
@dweiller I just ran into that in a different PR as well, seems to be a fluke.

dweiller commented:
The windows CI  failure seems to be the compiler running out of memory - not sure why this would happen, it shouldn't be cause by this PR.

kprotty commented:
A decent amount of the code could be simplified using comptime into something like:
```rs
// shared (dst, src)

copy(blk, offsets):
  for i in offsets: 
    dst[i..][0..blk].* = src[i..][0..blk].*

memcpy(len):
  if len == 0: return
  if len < 4: return copy(1, {0, len/2, len-1})

  v = max(4, suggestVectorSize(u8) or 0)
  inline for i in ctz(4)..ctz(v):
    if len <= 1 << (n+1): // @expect(false)
      return copy(1<<n, {0, len - (1<<n)})
      
  for i in 0..(len / v): copy(v, .{ i * v }) // 4-way auto-vec
  copy(v, .{ len - v })
```

But I think (at a higher level) we should go all the way: Facebook's folly [memcpy](<https://github.com/facebook/folly/blob/main/folly/memcpy.S>) is heavily optimized but also serves as their `memmove` since it's so similar. For sizes which cant be handled by a single `copy` above (anything after the `inline for`), they detect if src/dst overlap and switch to something similar but copied backwards.

Interestingly. they never use aligned loads since it doesn't seem worth it. Copy-forward does both load/store _unaligned_, but copy-backward uses aligned stores with `@prefetch(dst_to_write, .{ .rw = .write, .locality  = 3 })`. They also switch to `rep movsb` (it's like builtin CPU memcpy) if the conditions align for it to be fast (len >= 1024, pointers dont overlap, ERMSB cpuid enabled).

The general strategy of having memcpy & memmove be the same thing would be nice (perf & maintenance wise). Fast `memset` is also the same as fast memcpy but replace loads of src with a pre-`@splat` block/vector/register of the same byte.

dweiller commented:
> A decent amount of the code could be simplified using comptime into something like:

This looks pretty nice, I'll try it out.

> But I think (at a higher level) we should go all the way: Facebook's folly [memcpy](https://github.com/facebook/folly/blob/main/folly/memcpy.S) is heavily optimized but also serves as their `memmove` since it's so similar. For sizes which cant be handled by a single `copy` above (anything after the `inline for`), they detect if src/dst overlap and switch to something similar but copied backwards.

Our memcpy assumes that src/dst don't overlap, so a change like that would be a (non breaking, but significant) change in semantics that would affect performance.

> Interestingly. they never use aligned loads since it doesn't seem worth it. Copy-forward does both load/store _unaligned_, but copy-backward uses aligned stores with `@prefetch(dst_to_write, .{ .rw = .write, .locality = 3 })`. 

On my machine, IIRC, the aligned ops did affect performance, but this would also be something machine dependent. I have seen that the current wisdom seems to be that modern x86_64 doesn't really care, but what about other platforms?

> They also switch to `rep movsb` (it's like builtin CPU memcpy) if the conditions align for it to be fast (len >= 1024, pointers dont overlap, ERMSB cpuid enabled).

I did some research into `rep movsb` and read that the benefit and breakpoint for when it's beneficial is highly machine dependent, so I thought I'd leave using it for a future enhancement.

> Fast `memset` is also the same as fast memcpy but replace loads of src with a pre-`@splat` block/vector/register of the same byte.

I do actually have a memset branch locally as well - I could include it in this PR, but I haven't put as much work into it yet.

kprotty commented:
> Our memcpy assumes that src/dst don't overlap, so a change like that would be a (non breaking, but significant) change in semantics that would affect performance.

I rather mean to keep memcpy, the function with noalias and such, but have it just call memmove internally. This should keep the noalias optimizations applied by compiler replacing memcpy, but keep the vector-based / branch-optimized version at runtime.

> On my machine, IIRC, the aligned ops did affect performance

Was this in relation to the results above? I think the benchmark could report throughput in cycles/byte rather than ns. This is something used in benchmarks like `smhasher`; hash functions do similar reading/branching opts and take a small amount of time for short sizes. Combine this with not `clflush`ing dst/src and not cpu-core pinning, I'd assume ns-based time fluctuates more even with high iterations. 

I only mention this as aligned loads didn't seem to have an effect on [other benchmarks](<https://gist.github.com/Rexicon226/cb433c0a8639345e15224fd16cd66140>) so trying to somehow discover/rationalize the difference in the results. TBF, also haven't tested on anything outside avx2, avx512f, and apple_m1.

> rep movsb and read that the benefit and breakpoint for when it's beneficial is highly machine dependent

Indeed looks like it depends on micro-architecture specifically (being finnicky on zen2+). Seemed to be same speed as the normal vectorized loop at least on 5600x and 6900HS.

> I do actually have a memset branch locally as well - I could include it in this PR

Yea a separate PR would be best. Just wanted to mention their similarity.

dweiller commented:
> I rather mean to keep memcpy, the function with noalias and such, but have it just call memmove internally. This should keep the noalias optimizations applied by compiler replacing memcpy, but keep the vector-based / branch-optimized version at runtime.

Ah, okay - I understand now.

> > On my machine, IIRC, the aligned ops did affect performance
> 
> Was this in relation to the results above? I think the benchmark could report throughput in cycles/byte rather than ns. This is something used in benchmarks like `smhasher`; hash functions do similar reading/branching opts and take a small amount of time for short sizes. Combine this with not `clflush`ing dst/src and not cpu-core pinning, I'd assume ns-based time fluctuates more even with high iterations.

I think the benchmark has changed between when I tested (and the memcpy function has quite a bit too) - I'll re-check. Unless I'm missing something, I would be biased to keep the aligned strategy until/unless we diverge implementation based on whether a target has slow unaligned access or not, because I expect the impact on systems that don't care to be minimal (it costs one branch and one vector copy but this is only on large copies which are rare, and the effects of the increased code size), but systems that do have slower unaligned accesses will pay that cost proportional to the size of a long copy.

I can have the benchmark report cycles/byte, and do things like core-pinning and disabling frequency scaling next time I work on improving the benchmarking, but I think adding other benchmarks will probably take priority.

> I only mention this as aligned loads didn't seem to have an effect on [other benchmarks](https://gist.github.com/Rexicon226/cb433c0a8639345e15224fd16cd66140) so trying to somehow discover/rationalize the difference in the results. TBF, also haven't tested on anything outside avx2, avx512f, and apple_m1.

One reason, could be the architecture - looking at [Agner Fog's](https://www.agner.org/optimize/instruction_tables.pdf) tables my machine (Zen 1) has worse latencies/throughputs on unaligned intstructions in general.

andrewrk commented:
> Before merging it would be good to:

It looks like this is not ready for merging then? Can you close it and open a new one when it's ready for review & merging? Otherwise we can use the issue tracker to discuss ideas, plans, strategies, etc.

dweiller commented:
> > Before merging it would be good to:
> 
> It looks like this is not ready for merging then? Can you close it and open a new one when it's ready for review & merging? Otherwise we can use the issue tracker to discuss ideas, plans, strategies, etc.

I've just done some cleanup of commits and assuming it passes CI, I'm happy for it to be merged. The unchecked boxes are all things that are potential future work and could be broken out into new issues.

I would say the real question is what level of effort you want to see put into benchmarking before merging. I can certainly do more on my own - write more synthetic benchmarks, do some proper statistics, check some real-world impacts more carefully (I did see a small benefit in the zig compiler using the c backend, but that was a while ago and I don't remember the details), but I think any serious benchmarking effort would require help from others (in particular to check on more machines/architectures) and may be better done as part of [gotta-go-fast](https://github.com/ziglang/gotta-go-fast).

dweiller commented:
Looking at the table above again - would it be better to not include the optimisations done in this PR on `ReleaseSmall`? I'll double check the number are still accurate since the llvm-18 upgrade, but assuming they are is that too many extra bytes for ReleaseSmall?

elvircrn commented on lib/compiler_rt/memcpy.zig:
```diff
@@ -43,11 +43,13 @@ pub fn memcpy(noalias dest: ?[*]u8, noalias src: ?[*]const u8, len: usize) callc
-        if (len <= limit) {
-            memcpy_range2(limit / 2, dest.?, src.?, len);
-            return dest;
+    if (5 <= std.math.log2(2 * size)) {
```
Completely unncessary comment but perhaps you can simplify this to 16 <= size now. :)

Newbie zig question: is this log2 comptime evaluated anyway?

dweiller replied:
That log2 in the if statement wouldn't be evaluated at comptime (and it should be) - it should have a `comptime` keyword added. llvm was inlining/constant folding it so there wasn't actually a runtime test when using the llvm backend.

While `16 <= size` is simpler in isolation (and more so when the `comptime` is added, I think changing the test to compare to 16 makes it less clear why the test is needed.

dweiller commented:
> Looking at the table above again - would it be better to not include the optimisations done in this PR on `ReleaseSmall`? I'll double check the number are still accurate since the llvm-18 upgrade, but assuming they are is that too many extra bytes for ReleaseSmall?

The code size has changed minimally except for Debug, which now (presumably due to llvm-18) produces much more code (1745 bytes, as opposed to 143 bytes). In light of this I think it makes sense to revert to a simpler implementation like the original for debug and ReleaseSmall.

dweiller commented:
> But I think (at a higher level) we should go all the way: Facebook's folly [memcpy](https://github.com/facebook/folly/blob/main/folly/memcpy.S) is heavily optimized but also serves as their `memmove` since it's so similar. For sizes which cant be handled by a single `copy` above (anything after the `inline for`), they detect if src/dst overlap and switch to something similar but copied backwards.

Folly seems to do this in order to not break legacy code that (erroneously) depends on `memcpy` working for aliasing pointers. As far any Zig code goes, I don't think it makes sense to do this as `@memcpy` should have appropriate safety checks in safe builds.

Does Zig's compiler-rt implementation of `memcpy` ever get called from C code? If not I think there's no reason to follow Folly by making `memcpy` a `memmove`. If it does then it could be considered, but neither LLVM or GCC do it.

dweiller commented:
> Indeed looks like it depends on micro-architecture specifically (being finnicky on zen2+). Seemed to be same speed as the normal vectorized loop at least on 5600x and 6900HS.

There are two `CPUID` feature flags that may be a good indicator for when `rep movsb` is worth using, ERMSB (enhanced rep movsb) and FSRM (fast short rep mov). Unfortunately my CPU does not have these, so I cannot do any performance testing for it, but if someone with a CPU with ERMSB/FSRM wants to test you can try my `memcpy-rep-movsb` branch.

andrewrk commented:
I'm interested in landing this - I would like to see @Rexicon226 and @dweiller come to some kind of mutual understanding in order to proceed.

dweiller commented:
I've just run the distribution benchmark to compare the current PR with https://github.com/Rexicon226/zig/tree/memcpy-go-brr (commit [4e0ce4b](https://github.com/Rexicon226/zig/commit/4e0ce4b84d510d9b71a84e12ebb8a891ac0d15fb)) and got these results on my machine:

![comp](https://github.com/user-attachments/assets/fd3dd863-0443-4f9e-ad1d-a25a0caaf4a2)

I'm not sure what benchmark Rexicon was running in the past when he was comparing his version to old versions of this PR, so I can't check those results.

dweiller commented:
> A few other notes:
> 
>     * the copy loops are not manually unrolled (LLVM will unroll them on various targets).

I think that manually unrolling is not a good idea at present as LLVM further unrolls the copy loops on various targets and does so too aggressively. It would be reasonable to do manual unrolling if we have (`@loopHint`)[https://github.com/ziglang/zig/issues/22151]` since we do better unrolling via branchless tail cleanup.

andrewrk commented on lib/compiler_rt/memcpy.zig:
```diff
@@ -8,21 +9,283 @@ comptime {
 
-pub fn memcpy(noalias dest: ?[*]u8, noalias src: ?[*]const u8, len: usize) callconv(.C) ?[*]u8 {
-    @setRuntimeSafety(false);
+const CopyType = if (std.simd.suggestVectorLength(u8)) |vec_size|
```
You don't need to put "Type" in type names. It's already known to be a type, no need to also mention that it's a type in the type name.

A better name would be `Element`.

andrewrk commented on lib/compiler_rt/memcpy.zig:
```diff
@@ -8,21 +9,283 @@ comptime {
+else
+    memcpy_fast;
+
+fn memcpy_small(noalias dest: ?[*]u8, noalias src: ?[*]const u8, len: usize) callconv(.C) ?[*]u8 {
```
```zig
    @setRuntimeSafety(false);

    for (0..len) |i| {
        dest.?[i] = src.?[i];
    }

    return dest;
```

[generates better code](https://zig.godbolt.org/z/x5vGKq885)

also please follow [style guide](https://ziglang.org/documentation/master/#Style-Guide)

andrewrk replied:
this is the -OReleaseSmall version where the goal is to generate fewer bytes, so 2 bytes less code for idiomatic source is a good deal. if you wanted to hit the sweet spot instead you'd just compile compiler_rt with -OReleaseFast and the rest of the application with -OReleaseSmall. there's no niche for this sweet spot you're advocating for.

andrewrk commented on lib/compiler_rt/memcpy.zig:
```diff
@@ -8,21 +9,283 @@ comptime {
+    assert(std.math.isPowerOfTwo(copy_size));
+}
+
+pub const memcpy = if (builtin.mode == .ReleaseSmall)
+    memcpy_small
+else
+    memcpy_fast;
```
this logic belongs in the comptime block where it decides which function to export

dweiller commented on lib/compiler_rt/memcpy.zig:
```diff
@@ -8,21 +9,283 @@ comptime {
+fn memcpy_fast(noalias dest: ?[*]u8, noalias src: ?[*]const u8, len: usize) callconv(.C) ?[*]u8 {
+    @setRuntimeSafety(builtin.is_test);
+
+    const unroll_count = 1;
```
Since I settled on setting this to one it makes sense to remove, but I think that if we get #22151 in the future we may want it back, but that should probably come with measurements that show a difference between LLVM's unrolling and the branchless tail logic.

dweiller commented:
> I'm tired of waiting for @Rexicon226 so if you address this feedback I'll merge it. Thanks for working on this.

Should I also remove all the `inline` annotations from helper functions? They were originally added because without them (at least in some cases) LLVM would optimise into doing infinite recursive calls, but the new `-fno-builtin` behaviour should prevent this from happening. There shouldn't be any semantic reason to use `inline fn`, so the only consideration would be to ensure they don't get compiled to function calls. I assume release modes should always inline them, however, they do not get inlined if you compile in debug mode - is this an issue? I've got it in my head that compiler-rt is always compiled with `-OReleaseFast` or `-OReleaseSmall`, is that correct?

andrewrk commented:
For compiler_rt it's important that all the intrinsics are lowered to exactly one function. This helps codegen be better since the calls are known leaf calls, and it interacts better with `-ffunction-sections`.

andrewrk commented:
The commit f38d7a92cc1ca8059ba909a82f2f944966b52296 makes the memcpy implementation much more load-bearing, so it's important to rebase on top of latest master.

Can you update your perf measurements against latest master?

Can you take data points on aarch64 as well as x86_64?

dweiller commented:
> The commit [f38d7a9](https://github.com/ziglang/zig/commit/f38d7a92cc1ca8059ba909a82f2f944966b52296) makes the memcpy implementation much more load-bearing, so it's important to rebase on top of latest master.
> 
> Can you update your perf measurements against latest master?
> 
> Can you take data points on aarch64 as well as x86_64?

I'll update and re-measure, but I don't have a aarch64 machine to benchmark on, I'll need someone else willing to run the benchmarks.

RossComputerGuy commented:
> but I don't have a aarch64 machine to benchmark on, I'll need someone else willing to run the benchmarks.

I volunteer, I've got 2 aarch64 systems and even a RISC-V system that can be used.

dweiller commented:
I have rebased and updated the OP to reflect the current state of the PR. Note that this PR now (i.e. commit b7a887f0fb) splits `memmove` and `memcpy` implementations again, and also adds a `memmove` implementation for `ReleaseSamll` - if this PR doesn't get merged before the 0.14 release we will want to break that out into a separate change and merge that for 0.14.



> > but I don't have a aarch64 machine to benchmark on, I'll need someone else willing to run the benchmarks.
> 
> I volunteer, I've got 2 aarch64 systems and even a RISC-V system that can be used.

The benchmark I use in available at https://github.com/dweiller/zig-lib-bench if you can run the 'distrib' benchmark that would be great. If you have any issues feel free to raise issues there. It should be usable as is, but if you give me a few days I can try and make it a bit easier to use.

jacobly0 commented:
>     * For x86_64 machines that have FSRM it may be better to `rep movsb` for all/most cases (I don't have a suitable machine to test this)

Note that FSRM does not imply ERMSB.  FSRM just means that you can use `rep movsb` for a single preceding or trailing unaligned segment.

jacobly0 commented:
> I'll update and re-measure, but I don't have a aarch64 machine to benchmark on, I'll need someone else willing to run the benchmarks.

Between ZSF and I, we have at least skylake, zen2, zen3, zen4, neoverse_n1, and apple_a15.  If you provide specific instructions for producing a file for you to interpret and what cpus you need, I can do so.

andrewrk commented:
Thanks for the rebase and the re-writeup. I'm now happy with this as-is. ARM perf can be a followup investigation.

Appreciate your patience with this one, @dweiller.

dweiller commented:
I'm not sure what the CI failure on aarch64-windows is - as far as I see in the CI log there is just a exit 5 in the `f32_passed_to_variadic_fn` test case with no indicated cause. I don't know how variadic functions work under to hood - would it be using memcpy to pass the arguments?

andrewrk commented:
I don't have any working explanation for that failure at the moment, although unfortunately almost anything could be explained by the bug that #22541 is addressing.

With 216e0f37306d5cded92d547f65f9c2566d7b1523 landed, once that PR is also landed I think we should be looking at a significantly more reliable CI.

dweiller commented:
Ok - I'll just wait for that PR and rebase when it lands.

jacobly0 commented on lib/compiler_rt/memmove.zig:
```diff
@@ -0,0 +1,170 @@
+
+    while (@intFromPtr(s) > @intFromPtr(src + 128)) {
+        // zig fmt: off
+        const first  = @as(*align(1) const @Vector(32, u8), @ptrCast(s - 32)).*;
+        const second = @as(*align(1) const @Vector(32, u8), @ptrCast(s - 64)).*;
+        const third  = @as(*align(1) const @Vector(32, u8), @ptrCast(s - 96)).*;
+        const fourth = @as(*align(1) const @Vector(32, u8), @ptrCast(s - 128)).*;
+
+        @as(*align(32) @Vector(32, u8), @alignCast(@ptrCast(d - 32))).*  = first;
+        @as(*align(32) @Vector(32, u8), @alignCast(@ptrCast(d - 64))).*  = second;
+        @as(*align(32) @Vector(32, u8), @alignCast(@ptrCast(d - 96))).*  = third;
+        @as(*align(32) @Vector(32, u8), @alignCast(@ptrCast(d - 128))).* = fourth;
```
This code is not valid, as there is no guarantee that `@sizeOf(@Vector(32, u8)) == 32` for every target.  Perhaps you meant `[32]u8` instead of `@Vector(32, u8)` since `@sizeOf([32]u8) == 32` is guaranteed for every target.

dweiller replied:
That was the pre-existing memmove - I just moved it. I've got a branch with a memmove implementation based on the memcpy from this PR which wouldn't have that issue, but haven't had time to do benchmarking for it yet.

dweiller replied:
I suppose the `memcpy` implementation in this PR may have a similar issue if `std.simd.suggestVectorLength` does not return a suitable length.

andrewrk replied:
discussion continues at #22848

jacobly0 commented on lib/compiler_rt/memcpy.zig:
```diff
@@ -1,149 +1,213 @@
+    usize;
+
+comptime {
+    assert(@sizeOf(Element) >= @alignOf(Element));
+    assert(std.math.isPowerOfTwo(@sizeOf(Element)));
```
All types have a size that is a multiple of its alignment, therefore the first condition isn't checking for anything.  If you need a power of two, just use `std.math.floorPowerOfTwo` or something.

jacobly0 commented on lib/compiler_rt/memcpy.zig:
```diff
@@ -1,149 +1,213 @@
-fn memcpy(noalias opt_dest: ?[*]u8, noalias opt_src: ?[*]const u8, len: usize) callconv(.C) ?[*]u8 {
-    return memmove(opt_dest, opt_src, len);
+const Element = if (std.simd.suggestVectorLength(u8)) |vec_size|
+    @Type(.{ .vector = .{
+        .child = u8,
+        .len = vec_size,
+    } })
```
```suggestion
    @Vector(vec_size, u8)
```

jacobly0 commented on lib/compiler_rt/memcpy.zig:
```diff
@@ -1,149 +1,213 @@
+    const loop_count = max_bytes / @sizeOf(T);
+
+    for (dest[0..loop_count], src[0..loop_count]) |*d, s| {
+        d.* = s;
```
Yeah, same issue here, this is only required to copy `vec_len` bytes, but the vector size could be larger than that.

dweiller replied:
Verified to match what? I assume if I check that `vec_size == @Vector(vec_size, u8)` everything should be fine, and if it doesn't then either we can't use vectors or would have to do something like
```zig
const tmp: Element = byte_ptr_src[offset..][0..vec_size];
byte_ptr_dest[offset..][0..vec_size] = tmp;
```
in the copy loop.

jacobly0 replied:
Just a comptime assert should be fine, if `suggestVectorSize` doesn't return a vector with that property, it isn't making a very good suggestion.  The specific problem with the old `memmove` is that it just blindly uses a fixed vector without checking.

andrewrk commented on lib/compiler_rt/memcpy.zig:
```diff
@@ -1,149 +1,213 @@
+    const last = len - copy_len;
+    const pen = last - b;
+
+    dest[0..copy_len].* = src[0..copy_len].*;
+    dest[b..][0..copy_len].* = src[b..][0..copy_len].*;
+    dest[pen..][0..copy_len].* = src[pen..][0..copy_len].*;
+    dest[last..][0..copy_len].* = src[last..][0..copy_len].*;
```
These are recursive calls to memcpy. Somehow with optimizations on, that is avoided, but with `--debug-rt` the infinite loop is observed.

dweiller replied:
Can you tell me the full CLI invocation being used when this is observed? I had mitigations in the past to prevent LLVM making recursive memcpy calls, but changes to `-fno-builtin` should have prevented this from being possible.

andrewrk replied:
Sure, just pass `--debug-rt` to any CLI. For instance:

```
andy@bark ~/s/z/build-release (master)> stage3/bin/zig test ../test/behavior.zig --debug-rt
error: the following test command crashed:
/home/andy/src/zig/.zig-cache/o/d67352db25a508f21ac48aa4a6e52b3e/test --seed=0xa5e06631
```

andrewrk replied:
Discussion continues at #22848

dweiller replied:
> Because they don't seem aligned, and if they're not aligned then this _is_ a recursive call to memcpy.

I thought `-fno-builtin` is supposed to prevent calls to bultins from being generated isn't it?
